src/resolver.rs

```diff
@@ -19,27 +20,44 @@ pub trait Resolver {
 /// a thread to provide non-blocking address resolving.
 #[derive(Clone)]
 pub struct CpuPoolResolver {
-    pool: CpuPool,
+    pool: futures::executor::ThreadPool,
 }
 
 impl CpuPoolResolver {
     /// Create a new CpuPoolResolver with the given number of threads.
-    pub fn new(num_threads: usize) -> Self {
+    pub fn new(num_threads: NonZeroUsize) -> Self {
         CpuPoolResolver {
-            pool: CpuPool::new(num_threads),
+            pool: futures::executor::ThreadPool::builder()
+                .pool_size(num_threads.get())
+                .create()
+                .unwrap(),
         }
     }
 }
 
 impl Resolver for CpuPoolResolver {
     fn resolve(&self, host: &str) -> IoFuture<Vec<IpAddr>> {
         let host = format!("{}:0", host);
-        boxed(
-            self.pool
-                .spawn_fn(move || match host[..].to_socket_addrs() {
-                    Ok(it) => Ok(it.map(|s| s.ip()).collect()),
-                    Err(e) => Err(e),
-                }),
-        )
+        let mut pool = self.pool.clone();
+
+        async move {
+            let (tx, rx) = futures::channel::oneshot::channel();
+
+            pool.spawn_obj(
+                async move {
+                    tx.send(match host[..].to_socket_addrs() {
+                        Ok(it) => Ok(it.map(|s| s.ip()).collect()),
+                        Err(e) => Err(e),
+                    })
+                    .unwrap()
+                }
+                    .boxed()
+                    .into(),
+            )
+            .unwrap();
+
+            rx.await.unwrap()
```

Comment:
Are all of these unwraps panic free?

Reply:
Spawning on thread pool never fails. Replaced other unwraps.